### PR TITLE
Re-define amount, define tokens and token transfers, allow transfer and execute

### PR DESCRIPTION
**Pull request by andreibancioiu**

### CODE REVIEW

File: sdk-core/custom_tokens.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    extract_token_nonce_from_extended_identifier(extended_identifier: string): number;
+
+    // Given "FOO-abcdef-0a" returns "FOO-abcdef".
+    extract_token_identifier_from_extended_identifier(extended_identifier: string): string;
+
+    // Given "FOO-abcdef-0a" returns "FOO".
+    extract_token_name_from_extended_identifier(extended_identifier: string): string;
+
+    // Given "FOO-abcdef" returns "FOO".
+    extract_token_name_from_identifier(identifier: string): string;
```
Comment (popenta) [inline sdk-core/custom_tokens.md:29]: I understand the explicitly of having both methods, but maybe these can be only one method? Just a thought.

**Review by andrei-marinica — DISMISSED** [review]

File: sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md [hunk 2]
```diff
@@ -5,12 +5,15 @@ class SmartContractTransactionIntentsFactory:
...
     // Generally speaking, the constructor should be parametrized with a configuration object which defines entries such as:
     // "minGasLimit", "gasLimitPerByte" etc.
-    // The constructor should also be parametrized with a Codec instance, necessary to encode contract call arguments.
+    // The constructor may also be parametrized with a Codec instance, necessary to encode contract call arguments.
 
     create_transaction_intent_for_deploy({
         sender: IAddress;
         bytecode: bytes OR bytecodePath: Path;
         arguments: List[object] = [];
+        // Question for review: is this a valid case?
```
Comment (popenta) [inline sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md:14]: I've tested this with a contract that was payable on the `init` method, and it works. Indeed, this is a valid case. Same for the `upgrade` method.

File: sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md [hunk 3]
```diff
@@ -24,6 +27,8 @@ class SmartContractTransactionIntentsFactory:
         // If "function" is a reserved word in the implementing language, it should be replaced with a different name (e.g. "func" or "functionName").
         function: string;
         arguments: List[object] = [];
+        native_transfer_amount: Amount = 0;
+        custom_transfers: List[CustomTokenTransfer] = [];
```
Comment (popenta) [inline sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md:31]: 👍🏻

File: sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md [hunk 4]
```diff
@@ -32,6 +37,9 @@ class SmartContractTransactionIntentsFactory:
         contract: IAddress;
         bytecode: bytes OR bytecodePath: Path;
         arguments: List[object] = [];
+        // Question for review: is this a valid case?
+        native_transfer_amount: Amount = 0;
+        custom_transfers: List[CustomTokenTransfer] = [];
```
Comment (popenta) [inline sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md:42]: As stated above, this is a valid case.

File: sdk-core/custom_tokens.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    is_fungible_token(token: Token): boolean;
+
+    // Given "FOO-abcdef-0a" returns 10.
+    extract_token_nonce_from_extended_identifier(extended_identifier: string): number;
+
+    // Given "FOO-abcdef-0a" returns "FOO-abcdef".
+    extract_token_identifier_from_extended_identifier(extended_identifier: string): string;
+
+    // Given "FOO-abcdef-0a" returns "FOO".
+    extract_token_name_from_extended_identifier(extended_identifier: string): string;
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply sdk-core/custom_tokens.md]: Fixed.

File: sdk-core/custom_tokens.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply sdk-core/custom_tokens.md]: Removed, given @popenta's comment.

File: sdk-core/transaction-intents-factories/token_transfers_factory.md [hunk 6]
```diff
@@ -1 +1,27 @@
...
+
+    // Can throw:
+    // - ErrBadArguments
+    //
+    // All kinds of transfers (native and ESDT) should be handled.
+    // Bad usage should be reported.
+    create_transaction_intent_for_custom_transfers({
+        sender: IAddress;
+        receiver: IAddress;
+        custom_transfers: CustomTokenTransfer[];
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply sdk-core/transaction-intents-factories/token_transfers_factory.md:25]: Good point, that is correct. `native_transfer_amount` is not received as a parameter by this function.

File: sdk-core/custom_tokens.md (hunk 1 above)
Comment (andreibancioiu) [reply sdk-core/custom_tokens.md]: Good point, kept one method, updated comments.

File: sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md (hunk 2 above)
Comment (andreibancioiu) [reply sdk-core/transaction-intents-factories/smart_contract_transaction_intents_factory.md:14]: Good, thanks :tada: :raised_hands:

**Review by andrei-marinica — APPROVED** [review]

**Review by CiprianDraghici — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by popenta — APPROVED** [review]

**Review by ccorcoveanu — APPROVED** [review]

**Review by danielailie — APPROVED** [review]

File: sdk-core/amount.md [hunk 7]
```diff
@@ -1,21 +1,54 @@
...
-    num_decimals: number;
+    value: (big.Int|bigNumber|string)
+    to_string(): string
+```
+
+Within the specs, we use `Amount` and we mean `(bigNumber|string)`. Or, to put it differently, `(Go[big.Int]|Python[int]|JavaScript[BigNumber.Value|string])`.
+
+**Important:** 
+ - the value of an `Amount` is **always expressed in atomic units**. For example, to represent 1 EGLD, the value of `Amount` should be `1000000000000000000`, whether it's a Go `big.Int`, a Python `int`, a `string` etc.
+ - `Amount` **must not be concerned** with the number of decimals of the token. Just as the Protocol itself isn't concerned with this.
```
Comment (iulianpascalau) [inline sdk-core/amount.md:34]: :+1:

File: sdk-core/custom_tokens.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+```
+
+## TokenComputer
+
+```
+class CustomTokenComputer:
+    // The constructor is not captured by the specs; it's up to the implementing library to define it.
+
+    // Returns token.nonce == 0.
+    is_fungible_token(token: Token): boolean;
```
Comment (iulianpascalau) [inline sdk-core/custom_tokens.md:17]: do all methods described here can error? for example, what happens if I call the `extract_token_identifier_from_extended_identifier` with the `abcd` string?

File: sdk-core/transaction-intents-factories/token_transfers_factory.md [hunk 9]
```diff
@@ -1 +1,28 @@
...
+        native_transfer_amount: Amount;
+    }): TransactionIntent;
+
+    // Can throw:
+    // - ErrBadArguments
+    //
+    // All kinds of transfers (native and ESDT) should be handled.
+    // If multiple transfers are specified, a multi-ESDT transfer intent should be created.
+    // Bad usage should be reported.
+    create_transaction_intent_for_custom_transfers({
```
Comment (iulianpascalau) [inline sdk-core/transaction-intents-factories/token_transfers_factory.md:23]: question: how does the custom transfer & SC calls are handled? We call first this method and then work with the `TransactionIntent` to add the extra function & parameters? Do we have a factory for that so we should not go in details?

**Review by iulianpascalau — APPROVED** [review]

File: sdk-core/custom_tokens.md (hunk 8 above)
Comment (andreibancioiu) [reply sdk-core/custom_tokens.md:17]: Very good point! Will specify this in a separate PR :+1:

File: sdk-core/transaction-intents-factories/token_transfers_factory.md (hunk 9 above)
Comment (andreibancioiu) [reply sdk-core/transaction-intents-factories/token_transfers_factory.md:23]: For transfer & execute, the developer should use `SmartContractTransactionIntentsFactory` and pass the parameters (as needed):

```
native_transfer_amount: Amount = 0;
custom_transfers: List[CustomTokenTransfer] = [];
```

That factory is responsible with correctly composing the `tx.data`.

File: sdk-core/custom_tokens.md (hunk 8 above)
Comment (andreibancioiu) [reply sdk-core/custom_tokens.md:17]: Noted here: https://github.com/multiversx/mx-sdk-specs/issues/4.

**Pull request merged**